hazelcast/src/test/java/com/hazelcast/cache/TenantUnavailableTest.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache;
+
+import com.hazelcast.cache.impl.CacheProxy;
+import com.hazelcast.cache.impl.ICacheService;
+import com.hazelcast.config.CacheConfig;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.internal.util.ExceptionUtil;
+import com.hazelcast.partition.MigrationListener;
+import com.hazelcast.partition.MigrationState;
+import com.hazelcast.partition.ReplicaMigrationEvent;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import javax.cache.Cache;
+import javax.cache.CacheManager;
+import java.io.Serializable;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+
+import static com.hazelcast.cache.CachePartitionIteratorMigrationTest.putValuesToPartition;
+import static com.hazelcast.cache.CacheTestSupport.createServerCachingProvider;
+import static com.hazelcast.cache.CacheTestSupport.getCacheService;
+import static com.hazelcast.cache.HazelcastCacheManager.CACHE_MANAGER_PREFIX;
+import static com.hazelcast.cache.TenantControlTest.classesAlwaysAvailable;
+import static com.hazelcast.cache.TenantControlTest.destroyEventContext;
+import static com.hazelcast.cache.TenantControlTest.initState;
+import static com.hazelcast.cache.TenantControlTest.newConfig;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * App Server reload / app not loaded tests
+ *
+ * @author lprimak
+ */
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(QuickTest.class)
+public class TenantUnavailableTest extends HazelcastTestSupport {
+    private String cacheName;
+    private static final Set<String> disallowClassNames = new HashSet<>();
+    private static final CountDownLatch latch = new CountDownLatch(1);
+    private static boolean classLoadingFailed;
+
+    @Before
+    public void setup() {
+        cacheName = randomName();
+        classLoadingFailed = false;
+        initState();
+        classesAlwaysAvailable = false;
```

Comment:
`classesAlwaysAvailable` is false in all tests. Is there a test where it is true? Should we have a test for that?

Reply:
there is nothing to test that isn't rested already if that is true

Reply:
Then I'll remove this field as I don't see it's point.

Reply:
Ok, I think there was some use for this beforehand (it had to be set to true at one point) but with the current state it's no longer necessary